eval.go

```diff
@@ -75,23 +75,28 @@ func parseRemoteObject(v *runtime.RemoteObject, res interface{}) (err error) {
 		return
 	}
 
-	if v.Type == "undefined" || v.Value == nil {
+	value := v.Value
+	if value == nil {
+		// if js expression is undefined, it cannot resolve to any go type, so return ErrJSUndefined directly
+		if v.Type == "undefined" {
+			return ErrJSUndefined
+		}
```

Comment:
I don't think we should be able to distinguish what js code is undefined,undefined does not resolve to go type, so we should return the ErrJSUndefined error, and the caller can choose to ignore the error or check his js code

Reply:
`var m = new Map(); m.get('key')`  evaluate to undefined, so it return `ErrJSUndefined` is right

![image](https://user-images.githubusercontent.com/30597213/219532575-b5e19474-b038-4733-8fdd-67e581d21ff0.png)

It seems to be exactly what you said：
> I would assume that the user don't expect undefined/null at all. If it happens, we should report ErrJSUndefined or ErrJSNull to the user. If the user thinks that it's fine to receive undefined/null, then the user should write the code like this:
> 
> ```go
> var i *int
> err := chromedp.Run(ctx, chromedp.Evaluate("js expression that evaluates to integer/undefined/null", &i))
> ```
>

Reply:
> And if we want this, I think we should move this code block above value := v.Value.
I agree with you, but I'm not sure whether there is `v.Type=="undefined" && v.Value != nil` situation.
I assume that it exists then handle by json package.

Reply:
> It seems to be exactly what you said：
>> ... If the user thinks that it's fine to receive undefined/null, then the user should write the code like this:

It seems that you don't get my point. Imagine this scenario:

```go
var i *int
err := Run(ctx,
	Evaluate("some expression evaluated to undefined or null or an integer", &i),
	// other actions below
)
```

By saying "it's fine to receive undefined/null", I meant that even if the js result is `undefined`, __// other actions below__ should be executed.

This is not true if it returns `ErrJSUndefined` whenever the js result is `undefined`. And if forces me to change the code to:

```go
var i *int
err := Run(ctx,
	Evaluate("some expression evaluated to undefined or null or an integer", &i),
)

if err == nil || err == ErrJSUndefined {
	err = Run(ctx,
		// other actions below
	)
}
```

> but I'm not sure whether there is `v.Type=="undefined" && v.Value != nil` situation. I assume that it exists then handle by json package.

I think it's safe to say that `v.Value` is `nil` when `v.Type == "undefined"`.

Reply:
> If js expression is undefined, it cannot resolve to any go type, so return ErrJSUndefined directly

I think you meant this:

```go
var i int
err := json.Unmarshal([]byte("undefined"), &i)
// err: invalid character 'i' in literal null (expecting 'u'), i: 0
```

That's because "undefined" is not a valid json (see https://www.json.org/json-en.html). But that does not mean that `undefined` is not supported by the json package.

For example, I would say that `b` is `undefined` in the json `{"a":"valueA"}` (see the js code below).

```js
var obj = JSON.parse('{"a":"valueA"}');
console.log(typeof obj.b);
// undefined
```

And the json package can unmarshal `undefined` value:

```go
type UndefinedAsZeroValue struct {
	A string `json:"a"`
	B int    `json:"b"`
}
var obj1 UndefinedAsZeroValue
err := json.Unmarshal([]byte(`{"a":"valueA"}`), &obj1)
fmt.Printf("err: %v, obj1: %#v\n", err, obj1)
// err: <nil>, obj1: chromedp.UndefinedAsZeroValue{A:"valueA", B:0}

type UndefinedAsNil struct {
	A string `json:"a"`
	B *int   `json:"b"`
}
var obj2 UndefinedAsNil
err = json.Unmarshal([]byte(`{"a":"valueA"}`), &obj2)
fmt.Printf("err: %v, obj2: %#v\n", err, obj2)
// err: <nil>, obj2: chromedp.UndefinedAsNil{A:"valueA", B:(*int)(nil)}
```